Dotnet restore failed with MSB1001: Unknown switch

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

32.25.1

### Please select which platform you are using if self-hosting.

Azure DevOps (dev.azure.com)

### If you're self-hosting Renovate, tell us what version of the platform you run.

-

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

I just started to implement Renovate Bot for a project. The following error is occuring: "Command failed: dotnet restore RootProjectName/10 - Folder/ProjectName/ProjectName/ProjectName.csproj --force-evaluate --configfile /tmp/renovate/cache/others/nuget/randomName/nuget.config"
When trying to reproduce this locally with dotnet restore I noticed it is because of the " - " in the folder structure, for example when it is "RootProjectName/10-Folder/ProjectName/ProjectName/ProjectName.csproj" it works. Since it is a pretty big project it's a lot of work to rename everything.
Locally I could also fix it by putting the whole directory between quotes like this "dotnet restore "RootProjectName/10 - Folder/ProjectName/TestProjectName/TestProjectName.csproj" --force-evaluate --configfile /tmp/renovate/cache/others/nuget/randomName/nuget.config"

My suggestion would be to extend https://github.com/renovatebot/renovate/blob/1c0073d9a978c18b4a81ae89e1f4995897fdd434/lib/modules/manager/nuget/artifacts.ts at line 82 to include the quotes.



### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Failed to generate lock file (repository=Removed, branch=renovate/all-packages)
       "err": {
         "killed": false,
         "code": 1,
         "signal": null,
         "cmd": "dotnet restore RootProjectName/10 - Folder/ProjectName/ProjectName/ProjectName.csproj --force-evaluate --configfile /tmp/renovate/cache/others/nuget/randomName/nuget.config",
         "stdout": "MSBUILD : error MSB1001: Unknown switch.\nSwitch: -\n\nFor switch syntax, type \"MSBuild -help\"\n",
         "stderr": "",
         "message": "Command failed: dotnet restore RootProjectName/10 - Folder/ProjectName/ProjectName.csproj --force-evaluate --configfile /tmp/renovate/cache/others/nuget/randomName/nuget.config\n",
         "stack": "Error: Command failed: dotnet restore RootProjectName/10 - Folder/ProjectName/ProjectName/ProjectName.csproj --force-evaluate --configfile /tmp/renovate/cache/others/nuget/randomName/nuget.config\n\n    at ChildProcess.exithandler (node:child_process:399:12)\n    at ChildProcess.emit (node:events:526:28)\n    at ChildProcess.emit (node:domain:475:12)\n    at maybeClose (node:internal/child_process:1092:16)\n    at Process.ChildProcess._handle.onexit (node:internal/child_process:302:5)"
       }
```

</details>


### Have you created a minimal reproduction repository?

No reproduction repository

--- a/lib/modules/manager/nuget/artifacts.spec.ts
+++ b/lib/modules/manager/nuget/artifacts.spec.ts
@@ -1,19 +1,13 @@
-import { exec as _exec } from 'child_process';
 import { join } from 'upath';
-import { envMock, mockExecAll } from '../../../../test/exec-util';
-import { fs, git, mocked } from '../../../../test/util';
+import { envMock, exec, mockExecAll } from '../../../../test/exec-util';
+import { env, fs, git, mocked } from '../../../../test/util';
 import { GlobalConfig } from '../../../config/global';
 import type { RepoGlobalConfig } from '../../../config/types';
 import * as docker from '../../../util/exec/docker';
-import * as _env from '../../../util/exec/env';
 import * as _hostRules from '../../../util/host-rules';
 import type { UpdateArtifactsConfig } from '../types';
 import * as nuget from './artifacts';
-import {
-  getConfiguredRegistries as _getConfiguredRegistries,
-  getDefaultRegistries as _getDefaultRegistries,
-  getRandomString as _getRandomString,
-} from './util';
+import * as util from './util';
 
 jest.mock('child_process');
 jest.mock('../../../util/exec/env');
@@ -22,14 +16,8 @@ jest.mock('../../../util/host-rules');
 jest.mock('../../../util/git');
 jest.mock('./util');
 
-const exec: jest.Mock<typeof _exec> = _exec as any;
-const env = mocked(_env);
-const getConfiguredRegistries: jest.Mock<typeof _getConfiguredRegistries> =
-  _getConfiguredRegistries as any;
-const getDefaultRegistries: jest.Mock<typeof _getDefaultRegistries> =
-  _getDefaultRegistries as any;
-const getRandomString: jest.Mock<typeof _getRandomString> =
-  _getRandomString as any;
+const { getConfiguredRegistries, getDefaultRegistries, getRandomString } =
+  mocked(util);
 const hostRules = mocked(_hostRules);
 
 const adminConfig: RepoGlobalConfig = {
@@ -44,13 +32,13 @@ describe('modules/manager/nuget/artifacts', () => {
   beforeEach(() => {
     jest.resetAllMocks();
     jest.resetModules();
-    getDefaultRegistries.mockReturnValue([] as any);
+    getDefaultRegistries.mockReturnValue([]);
     env.getChildProcessEnv.mockReturnValue(envMock.basic);
     fs.ensureCacheDir.mockImplementation((dirName: string) =>
       Promise.resolve(`others/${dirName}`)
     );
     git.getFileList.mockResolvedValueOnce([]);
-    getRandomString.mockReturnValue('not-so-random' as any);
+    getRandomString.mockReturnValue('not-so-random');
     GlobalConfig.set(adminConfig);
     docker.resetPrefetchedImages();
   });
@@ -70,17 +58,19 @@ describe('modules/manager/nuget/artifacts', () => {
         config,
       })
     ).toBeNull();
-    expect(execSnapshots).toMatchSnapshot();
+    expect(execSnapshots).toBeEmptyArray();
   });
 
   it('aborts if lock file is unchanged', async () => {
     const execSnapshots = mockExecAll(exec);
-    fs.getSiblingFileName.mockReturnValueOnce('packages.lock.json');
-    fs.readLocalFile.mockResolvedValueOnce('Current packages.lock.json' as any);
-    fs.readLocalFile.mockResolvedValueOnce('Current packages.lock.json' as any);
+    fs.getSiblingFileName.mockReturnValueOnce(
+      'path/with space/packages.lock.json'
+    );
+    fs.readLocalFile.mockResolvedValueOnce('Current packages.lock.json');
+    fs.readLocalFile.mockResolvedValueOnce('Current packages.lock.json');
     expect(
       await nuget.updateArtifacts({
-        packageFileName: 'project.csproj',
+        packageFileName: 'path/with space/project.csproj',
         updatedDeps: [{ depName: 'foo' }, { depName: 'bar' }],
         newPackageFileContent: '{}',
         config,
@@ -92,8 +82,8 @@ describe('modules/manager/nuget/artifacts', () => {
   it('updates lock file', async () => {
     const execSnapshots = mockExecAll(exec);
     fs.getSiblingFileName.mockReturnValueOnce('packages.lock.json');
-    fs.readLocalFile.mockResolvedValueOnce('Current packages.lock.json' as any);
-    fs.readLocalFile.mockResolvedValueOnce('New packages.lock.json' as any);
+    fs.readLocalFile.mockResolvedValueOnce('Current packages.lock.json');
+    fs.readLocalFile.mockResolvedValueOnce('New packages.lock.json');
     expect(
       await nuget.updateArtifacts({
         packageFileName: 'project.csproj',
@@ -108,8 +98,8 @@ describe('modules/manager/nuget/artifacts', () => {
   it('does not update lock file when non-proj file is changed', async () => {
     const execSnapshots = mockExecAll(exec);
     fs.getSiblingFileName.mockReturnValueOnce('packages.lock.json');
-    fs.readLocalFile.mockResolvedValueOnce('Current packages.lock.json' as any);
-    fs.readLocalFile.mockResolvedValueOnce('New packages.lock.json' as any);
+    fs.readLocalFile.mockResolvedValueOnce('Current packages.lock.json');
+    fs.readLocalFile.mockResolvedValueOnce('New packages.lock.json');
     expect(
       await nuget.updateArtifacts({
         packageFileName: 'otherfile.props',
@@ -118,14 +108,14 @@ describe('modules/manager/nuget/artifacts', () => {
         config,
       })
     ).toBeNull();
-    expect(execSnapshots).toMatchSnapshot();
+    expect(execSnapshots).toBeEmptyArray();
   });
 
   it('does not update lock file when no deps changed', async () => {
     const execSnapshots = mockExecAll(exec);
     fs.getSiblingFileName.mockReturnValueOnce('packages.lock.json');
-    fs.readLocalFile.mockResolvedValueOnce('Current packages.lock.json' as any);
-    fs.readLocalFile.mockResolvedValueOnce('New packages.lock.json' as any);
+    fs.readLocalFile.mockResolvedValueOnce('Current packages.lock.json');
+    fs.readLocalFile.mockResolvedValueOnce('New packages.lock.json');
     expect(
       await nuget.updateArtifacts({
         packageFileName: 'project.csproj',
@@ -134,14 +124,14 @@ describe('modules/manager/nuget/artifacts', () => {
         config,
       })
     ).toBeNull();
-    expect(execSnapshots).toMatchSnapshot();
+    expect(execSnapshots).toBeEmptyArray();
   });
 
   it('performs lock file maintenance', async () => {
     const execSnapshots = mockExecAll(exec);
     fs.getSiblingFileName.mockReturnValueOnce('packages.lock.json');
-    fs.readLocalFile.mockResolvedValueOnce('Current packages.lock.json' as any);
-    fs.readLocalFile.mockResolvedValueOnce('New packages.lock.json' as any);
+    fs.readLocalFile.mockResolvedValueOnce('Current packages.lock.json');
+    fs.readLocalFile.mockResolvedValueOnce('New packages.lock.json');
     expect(
       await nuget.updateArtifacts({
         packageFileName: 'project.csproj',
@@ -160,8 +150,8 @@ describe('modules/manager/nuget/artifacts', () => {
     GlobalConfig.set({ ...adminConfig, binarySource: 'docker' });
     const execSnapshots = mockExecAll(exec);
     fs.getSiblingFileName.mockReturnValueOnce('packages.lock.json');
-    fs.readLocalFile.mockResolvedValueOnce('Current packages.lock.json' as any);
-    fs.readLocalFile.mockResolvedValueOnce('New packages.lock.json' as any);
+    fs.readLocalFile.mockResolvedValueOnce('Current packages.lock.json');
+    fs.readLocalFile.mockResolvedValueOnce('New packages.lock.json');
     expect(
       await nuget.updateArtifacts({
         packageFileName: 'project.csproj',
@@ -177,8 +167,8 @@ describe('modules/manager/nuget/artifacts', () => {
     GlobalConfig.set({ ...adminConfig, binarySource: 'global' });
     const execSnapshots = mockExecAll(exec);
     fs.getSiblingFileName.mockReturnValueOnce('packages.lock.json');
-    fs.readLocalFile.mockResolvedValueOnce('Current packages.lock.json' as any);
-    fs.readLocalFile.mockResolvedValueOnce('New packages.lock.json' as any);
+    fs.readLocalFile.mockResolvedValueOnce('Current packages.lock.json');
+    fs.readLocalFile.mockResolvedValueOnce('New packages.lock.json');
     expect(
       await nuget.updateArtifacts({
         packageFileName: 'project.csproj',
@@ -192,7 +182,7 @@ describe('modules/manager/nuget/artifacts', () => {
 
   it('catches errors', async () => {
     fs.getSiblingFileName.mockReturnValueOnce('packages.lock.json');
-    fs.readLocalFile.mockResolvedValueOnce('Current packages.lock.json' as any);
+    fs.readLocalFile.mockResolvedValueOnce('Current packages.lock.json');
     fs.writeLocalFile.mockImplementationOnce(() => {
       throw new Error('not found');
     });
@@ -216,8 +206,8 @@ describe('modules/manager/nuget/artifacts', () => {
   it('authenticates at registries', async () => {
     const execSnapshots = mockExecAll(exec);
     fs.getSiblingFileName.mockReturnValueOnce('packages.lock.json');
-    fs.readLocalFile.mockResolvedValueOnce('Current packages.lock.json' as any);
-    fs.readLocalFile.mockResolvedValueOnce('New packages.lock.json' as any);
+    fs.readLocalFile.mockResolvedValueOnce('Current packages.lock.json');
+    fs.readLocalFile.mockResolvedValueOnce('New packages.lock.json');
     getConfiguredRegistries.mockResolvedValueOnce([
       {
         name: 'myRegistry',
@@ -250,8 +240,8 @@ describe('modules/manager/nuget/artifacts', () => {
   it('strips protocol version from feed url', async () => {
     const execSnapshots = mockExecAll(exec);
     fs.getSiblingFileName.mockReturnValueOnce('packages.lock.json');
-    fs.readLocalFile.mockResolvedValueOnce('Current packages.lock.json' as any);
-    fs.readLocalFile.mockResolvedValueOnce('New packages.lock.json' as any);
+    fs.readLocalFile.mockResolvedValueOnce('Current packages.lock.json');
+    fs.readLocalFile.mockResolvedValueOnce('New packages.lock.json');
     getConfiguredRegistries.mockResolvedValueOnce([
       {
         name: 'myRegistry',

--- a/lib/modules/manager/nuget/artifacts.ts
+++ b/lib/modules/manager/nuget/artifacts.ts
@@ -42,14 +42,18 @@ async function addSourceCmds(
       url: registry.url,
     });
     const registryInfo = parseRegistryUrl(registry.url);
-    let addSourceCmd = `dotnet nuget add source ${registryInfo.feedUrl} --configfile ${nugetConfigFile}`;
+    let addSourceCmd = `dotnet nuget add source ${quote(
+      registryInfo.feedUrl
+    )} --configfile ${quote(nugetConfigFile)}`;
     if (registry.name) {
       // Add name for registry, if known.
       addSourceCmd += ` --name ${quote(registry.name)}`;
     }
     if (username && password) {
       // Add registry credentials from host rules, if configured.
-      addSourceCmd += ` --username ${username} --password ${password} --store-password-in-clear-text`;
+      addSourceCmd += ` --username ${quote(username)} --password ${quote(
+        password
+      )} --store-password-in-clear-text`;
     }
     result.push(addSourceCmd);
   }
@@ -78,8 +82,10 @@ async function runDotnetRestore(
   const cmds = [
     ...(await addSourceCmds(packageFileName, config, nugetConfigFile)),
     ...dependentPackageFileNames.map(
-      (f) =>
-        `dotnet restore ${f} --force-evaluate --configfile ${nugetConfigFile}`
+      (fileName) =>
+        `dotnet restore ${quote(
+          fileName
+        )} --force-evaluate --configfile ${quote(nugetConfigFile)}`
     ),
   ];
   await exec(cmds, execOptions);

